### PR TITLE
[WIP] Code generation template for new coins.

### PR DESCRIPTION
**Pull request by optout21**
## Description

Add code generation template, for the case when adding a new coin.  It generates a few implementation and test files.

## Testing instructions

Add a new entry to coins.json.
Invoke codegen/bin/newcoin, with the id of the coin.
Check resulting files.

## Types of changes

* New feature; New code generation script.

## Checklist

- [ ] Prefix PR title with `[WIP]` if necessary.
- [ ] Add tests to cover changes as needed.
- [ ] Update documentation about new coin addition!

### CODE REVIEW

File: codegen/bin/newcoin [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  format_name(coin).downcase
+end
+
+command_line_args = ARGV
+if command_line_args.length < 1
+    puts "Usage: newcoin <new_coin_id>"
+    return
+end
+
+def self.generateFile(templateFile, folder, fileName, coin)
```
Comment (hewigovens) [inline codegen/bin/newcoin]:
```suggestion
def self.generate_file(templateFile, folder, fileName, coin)
```

File: codegen/bin/cointests [hunk 2]
```diff
@@ -36,6 +36,19 @@ def self.coin_type(path)
   path.split('/')[2].chomp("'")
 end
 
+def self.generateCoinTestFile(coin, templateFile)
```
Comment (hewigovens) [inline codegen/bin/cointests]:
```suggestion
def self.generate_coin_testfile(coin, templateFile)
```

File: codegen/bin/newcoin [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+generator = CodeGenerator.new(entities: [entity], files: [file], output_folder: ".")
+
+@coins = coins
+
+coin_test_gen = CoinTestGen.new()
+
+coins.each do |coin|
+    if coin['id'] == coin_id
+        generateFile("newcoin/Address.h.erb", "src/#{format_name(coin)}", "Address.h", coin)
+        generateFile("newcoin/Address.cpp.erb", "src/#{format_name(coin)}", "Address.cpp", coin)
```
Comment (hewigovens) [inline codegen/bin/newcoin]: also add `Signer` and `TWSigner`?

File: codegen/lib/templates/newcoin/Address.h.erb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    explicit Address(const std::string& string);
+
+    /// Initializes a <%= format_name(coin) %> address with a public key.
+    explicit Address(const PublicKey& publicKey);
+
+    /// Returns a string representation of the address.
+    std::string string() const;
+
+    /// Returns the binary representation of the address (key hash) 
+    const Data keyHash() const;
```
Comment (hewigovens) [inline codegen/lib/templates/newcoin/Address.h.erb]: let's exclude this, not every address needs it

File: codegen/lib/templates/newcoin/AddressTests.cpp.erb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+TEST(<%= format_name(coin) %>Address, Invalid) {
+    ASSERT_FALSE(Address::isValid("__ADD_INVALID_ADDRESS_HERE__"));
+
+    // TODO: Add more tests
+}
+
+TEST(<%= format_name(coin) %>Address, FromPrivateKey) {
+    // TODO: Check public key type, finalize implementation
+
+    auto privateKey = PrivateKey(parse_hex("526d96fffdbfe787b2f00586298538f9a019e97f6587964dc61aae9ad1d7fa23"));
```
Comment (hewigovens) [inline codegen/lib/templates/newcoin/AddressTests.cpp.erb]: remove test key? just use comment

File: codegen/lib/templates/newcoin/AddressTests.cpp.erb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    auto address = Address(privateKey.getPublicKey(TWPublicKeyTypeED25519));
+    ASSERT_EQ(address.string(), "__ADD_RESULTING_ADDRESS_HERE__");
+}
+
+TEST(<%= format_name(coin) %>Address, FromPublicKey) {
+    // TODO: Check public key type, finalize implementation
+    
+    auto publicKey = PublicKey(parse_hex("c2b423afa8b0095e5ae105668b91b2132db4dadbf38acfc64908d3476a00191f"), TWPublicKeyTypeED25519);
```
Comment (hewigovens) [inline codegen/lib/templates/newcoin/AddressTests.cpp.erb]: remove test key? just use comment

File: tests/Wanchain/TWCoinTypeTests.cpp [hunk 7]
```diff
@@ -15,7 +15,7 @@
 
 TEST(TWWanchainCoinType, TWCoinType) {
     auto symbol = WRAPS(TWCoinTypeConfigurationGetSymbol(TWCoinTypeWanchain));
-    auto txId = TWStringCreateWithUTF8Bytes("0x180ea96a3218b82b9b35d796823266d8a425c182507adfe5eeffc96e6a14d856");
```
Comment (hewigovens) [inline tests/Wanchain/TWCoinTypeTests.cpp]: don't override real world tx id

**Review by hewigovens — APPROVED** [review]

**Review by vikmeup — APPROVED** [review]

**Pull request merged**